contracts/FeralfileArtworkV4_2.sol

```diff
@@ -0,0 +1,339 @@
+    error MintNotEnabled();
+    error FunctionNotSupported();
+    error EmptyURI();
+
+    struct TokenInfo {
+        string imageURI;
+        bytes parameters;
+        uint256 coinAmount;
+    }
+
+    struct MintDataWithIndex {
+        uint256 seriesId;
+        uint256 tokenId;
+        address owner;
+        uint256 tokenIndex;
+    }
+
+    FeralfileToken private _erc20Token;
+
+    mapping(uint256 => TokenInfo) private _tokenInfos; // tokenID => tokenInfo
+    mapping(uint256 => uint256) private _tokenIndexes; // tokenID => tokenIndex
+
+    string public tokenPlaceholderAnimationURI;
+    string public tokenPlaceholderImageURI;
+
+    string public artworkFileURI;
+
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        bool burnable_,
+        bool bridgeable_,
+        address signer_,
+        address vault_,
+        address costReceiver_,
+        string memory contractURI_,
+        uint256[] memory seriesIds_,
+        uint256[] memory seriesMaxSupplies_,
+        address erc20ContractAddress_
+    )
+        FeralfileExhibitionV4_1(
+            name_,
+            symbol_,
+            burnable_,
+            bridgeable_,
+            signer_,
+            vault_,
+            costReceiver_,
+            contractURI_,
+            seriesIds_,
+            seriesMaxSupplies_
+        )
+    {
+        _erc20Token = FeralfileToken(erc20ContractAddress_);
+    }
+    
+    /// @notice Mint new collection of Artwork
+    /// @dev the function iterates over the array of MintDataWithIndex to call the internal function _mintArtwork
+    /// @param data an array of MintDataWithIndex
+    function mintArtworksWithIndex(
+        MintDataWithIndex[] calldata data
+    ) external onlyAuthorized {
+        if (!mintable) {
+            revert MintNotEnabled();
+        }
+
+        for (uint256 i = 0; i < data.length; i++) {
+            _tokenIndexes[data[i].tokenId] = data[i].tokenIndex;
+            _mintArtwork(data[i].seriesId, data[i].tokenId, data[i].owner);
+        }
+    }
+
+    /// @notice override revert mint new collection of Artwork
+    function mintArtworks(MintData[] calldata) external override view onlyAuthorized {
+        revert FunctionNotSupported();
+    }
+
+    /// @notice Update the imageURI & parameters of an edition to a new value
+    function updateTokenInformation(uint256 tokenId, string calldata imageURI, bytes calldata parameters, uint256 coinAmount)
+        external
+        onlyAuthorized
+    {
+        if (!_exists(tokenId)) {
+            revert TokenIDNotFound();
+        }
+
+        TokenInfo memory tokenInfo = _tokenInfos[tokenId];
+        _tokenInfos[tokenId] = TokenInfo(imageURI, parameters, coinAmount);
+
+        // mint token for the owner
+        uint256 mintAmount = coinAmount - tokenInfo.coinAmount;
+        if (mintAmount > 0) {
+            address owner = ownerOf(tokenId);
+            _erc20Token.mint(owner, mintAmount * 10**18);
```

Comment:
Should call `decimal()` from token contract instead of hard-coding the `18` decimals.